docs/programmers_guide/witness_formal_spec.md

```diff
@@ -0,0 +1,108 @@
+# Block Witness Formal Specification
+
+Block witness uses 2 stacks:
+- node stack: `NODES`
+- hashes stack: `HASHES`
+
+That allows to calculate trie hashes w/o building the tries themselves.
+
+## Instructions
+
+### Substitution rules
+
+a typical substitution rule shows what values on stacks are replaced (before an INSTRUCTION), to by which values (after an INSTRUCTION)
+
+```
+NODECONST(nodeValue), NODECONST(nodeValue2)   // pop nodes on the NODES stack
+HASHCONST(hashValue),HASHCONST(hashValue2)    // pop nodes on the HASHES stack
+INSTRUCTION(params) |=>                       // the current instruction with params
+NODECONST(helperFunc(nodeValue, nodeValue2, params))    // push node(s) to the NODES stack
+HASHCONST(heperFunc2(hashValue, hashValue2, params))    // push hash(es) to the HASHES stack
+```
+
+### `BRANCH mask`
+
+This instruction pops `NBITSET(mask)` items from both node stack and hash stack (up to 16 for each one). Then it pushes a new branch node on the node stack that has children according to the stack; it also pushes a new hash to the hash stack.
+
+**Substitution rules**
+```
+BRANCH(mask) |=>
+NODECONST(branchNode())
+HASHCONST(KECCAK())
+
+---
+
+NODECONST(n0)
+HASHCONST(h0)
+BRANCH(mask) |=>
+NODECONST(branchNode(MAKE_VALUES_ARRAY(mask, n0))
+HASHCONST(keccak(CONCAT(MAKE_VALUES_ARRAY(mask, h0)))
+
+---
+
+NODECONST(n0) NODECONST(n1)
```

Comment:
There should be no branch nodes with fewer than two children, so perhaps this needs to be the first one

Reply:
I would also organise things on the stack as NODECONST HASHCONST NODECONST HASHCONST, rather than group of NODECONSTs and then group of HASHCONSTs, otherwise you would also need rules to reorder them